### DISCUSSION TITLE
Builder tutorial 4 - event-type list displays NIL

### DISCUSSION THREAD

**Discussion by aykaramba** [rabbibotton/clog#161] [General]
Hi.

I an re-doing the Tutorial 4 Database App using the latest CLOG cloned last night.  I have successfully setup the config-event-types panel for the app and everything works except the event-type list text box.  Instead of showing the row contents it shows NIL.

Details:

1) All buttons work, I can add a new item, edit an item or delete the item.  The text box will even update normally but instead of showing the content of rowid it just shows nil.

2) Screenshot of the event_type table using an sqlite editor, you can see entries in there, they add / remove properly:

![db](https://user-images.githubusercontent.com/16431651/177587016-118ef887-7033-4807-9a47-a2ca4df9c63d.png)

3) Screenshot of the pop edit dialog, it lets me edit the data and will save the actual content of the form correct as it is typed in, however, it displays NIL as per this screenshot:

![config-event-type](https://user-images.githubusercontent.com/16431651/177587268-cff2f3c3-2df7-4d6b-9b8b-da17c2d86d6b.png)

4) I have set up the event-type-list text box as per the tutorial description.  Obviously the cc-db database control is connecting properly because it is writing and editing the data, the event-type-list is AFTER the database control (obviously) and I have setup the value field and value display field as per the tutorial here:

![builder](https://user-images.githubusercontent.com/16431651/177587600-8e692ee4-d548-4a62-bf80-499e4e9bad4d.png)

It feels like I have something misconfigured somewhere, as simple as having the wrong name set somewhere in builder.  I have tried changing values for value field and value display field as well as the table name and the text field responds correctly if that data is incorrect, namely, it displays no information at all.

Has anyone run into this?  Anyone know why this might be happening or have suggestions on how to fix it?

Thanks.

**Comment by rabbibotton** [comment]
![Screen Shot 2022-07-06 at 1 01 52 PM](https://user-images.githubusercontent.com/11616789/177604696-2bd0f4b9-bdf7-4232-86d1-c6e11d46518e.png)

**Comment by rabbibotton** [comment]
Revisiting this though shows me some areas I need to work on, like being able to set the database location outside the individual panels.

**Reply by aykaramba** [reply]
Thank you kindly, that works.

I will update the tutorial to just add a bit of explanation to that one paragraph around page 11 for clarification.  Updates later today.

Thanks Rabbi.

**Comment by rabbibotton** [comment]
I modified the clos-contact repo so that the database names are set to a global *database* that is defined in clos-contact.lisp based on the projects location. For now if you set in a database control the database name to start with a * it assumes the database name is a variable local to the panels package.

**Reply by rabbibotton** [reply]
<img width="1034" alt="Screen Shot 2022-07-06 at 6 30 28 PM" src="https://user-images.githubusercontent.com/11616789/177654154-ce27289b-e207-4e34-ac44-1d89a5dbe5ff.png">
<img width="813" alt="Screen Shot 2022-07-06 at 6 30 44 PM" src="https://user-images.githubusercontent.com/11616789/177654156-f4d24976-1313-44c5-9796-5cc1bd34f2b8.png">

**Reply by aykaramba** [reply]
Oh neat!  Thanks.

**Comment by aykaramba** [comment]
Hi guys.

REQUEST: I am looking for some assistance troubleshooting the CLOG Builer Tutorial 4 project file from here: https://docs.google.com/document/d/185Sh9rMhSCEwt14chbn4_gi0CV6xwWs8Z59rUvEWCN8/edit

PROBLEM: I have tried to rebuild the app from the tutorial a number of times and although each time I have made errors along the way, I cannot get over a few common and repeatable (repeatable by me I mean) errors.  Basically, everything seems to work correctly except:

* Drop down table lookup - Any time I have a dropdown table lookup field, even if I have tripple checked the tutorial instructions they do not populate with data.  

* The find, next, update and delete buttons do not work and generate errors (INSERT WORKS!) -  In the Manage Contacts panel, the INSERT button works perfectly (I can see the entries in the sqlite db) but the other buttons generate errors.

DETAILS:

Here is a screenshot for visual reference:

![Untitled](https://user-images.githubusercontent.com/16431651/178106389-a3ddc910-9eb9-4a8e-a093-abe4842966be.png)

Description:

#1 - All of CONFIG EVENT TYPES works perfectly.

#2 - MANAGE CONTACTS seems to actually work because INSERT does insert the form data into the sqlite db.  However, all of the other buttons generate query errors that I don't fully understand.  It feels like I have done something wrong, but, I have done this a few times now and I am pretty sure I have followed the tutorial instructions fully.

#3 - The CONTACT EVENTS panel works fully EXCEPT drop down table lookup fields.  As before, I have done this a few times now and I cannot see what I am getting wrong relative to the tutorial instructions.

#4 - While the REPORT CONTACTS panel works great, the REPORT EVENTS panel does not.  Again, the drop down table lookup field doesn't seem to work.

MY CODE IS HERE:  https://github.com/aykaramba/clog-contacts-tutorial

MY PROJECT / CODE DESCRIPTION: I have done this tutorial a few times now from scratch in various ways, including separate panels as per tutorial and various combinations to try and figure out what I am doing wrong.  In the uploaded code I decided to stuff all of the panels into a SINGLE PANEL to force my self to think through the examples and figure out what I am doing wrong.  Yes, I am using a DATABASE CONTROL for each section even though I can / should only use it once.  The reason that I decide to repeat a DATABASE CONTROL for each section is to force my self to rename key components and test how everything is hooked up and related to each other instead of just copy and paste.  Also, the layouts are ugly on purpose, I am focusing on just getting the code to work first, making it look nice is easy.

Overall, this tutorial is really great and it has taught me a great deal about how to build apps in CLOG and using CLOG Builder.  I am getting used to builder and it is a very useful tools.  The performance speedups in the latest release REALLY make me see that  it possible build apps, especially after doing Tutorial 4.

Would anyone be able to take a quick peek at my tutorial code and give me some hints as to what I might be doing wrong?  Anyone recently build tutorial 4 successfully?  Any tips?

Thanks!

**Comment by aykaramba** [comment]
Or, if someone has a working copy of the tutorial they can upload for study, that would be helpful as well.

Thanks!

**Comment by rabbibotton** [comment]
Did you compare it to the working copy I put up :)

https://github.com/rabbibotton/clos-contact

**Comment by aykaramba** [comment]
Oh, I missed this!  Thank you kindly.

**Comment by aykaramba** [comment]
UPDATE:  Everything works!  

Thank you so much for providing a working sample, I made two tiny errors.  I have to do a bit more work with CLOG but I will contribute just a bit of editing to some of the tutorials to assist readers with the process.   As a side note, I love how you write your tutorials, there is so much personality there.  Please never change or change your tutorial writing style, it is perfect.  We can all help just touch up details here and there to punch it up.

I have to ask you a few questions about the manage contacts panel:

1) How does the Find function work ... how do you search for a name?  Searching for David david david* *david* david% results in errors similar to this:  

DB Error: no such column: david (Code: Error)

or

DB Error: neaer "order": syntax error (Code: Error)

The errors are from running your clos-contact code.

2) I assume that once Find returns data that Next / Insert / Update / Delete should work?  Right now they return similar errors as well.

Thank you kindly for your amazing work.  This tool is so much fun.
